File: kotlin/src/main/kotlin/leetcode/31.kt
package leetcode

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class NextPermutation31 {
    fun nextPermutation(nums: IntArray): Unit {
        val n = nums.size
        if (n <= 1) return

        fun swap(i: Int, j: Int) {
            nums[i] = nums[j].also { nums[j] = nums[i] }
        }

        fun rev(i: Int, j: Int) {
            for (k in 0 until (j - i + 1) / 2)
                swap(i + k, j - k)
        }

        val low = (n - 2 downTo 0).firstOrNull { nums[it] < nums[it + 1] }
        if (low == null) {
            rev(0, n - 1)
        } else {
            val high = (n - 1 downTo low + 1).first { nums[it] > nums[low] }
            swap(low, high)
            rev(low + 1, n - 1)
        }
    }

    @Test
    fun test1() {
        var nums = parseIntArray("[1,2,3]")
        nextPermutation(nums)
        assertEquals(parseIntArray("[1,3,2]").toList(), nums.toList())

        nums = parseIntArray("[3,2,1]")
        nextPermutation(nums)
        assertEquals(parseIntArray("[1,2,3]").toList(), nums.toList())

        nums = parseIntArray("[1,1,5]")
        nextPermutation(nums)
        assertEquals(parseIntArray("[1,5,1]").toList(), nums.toList())
    }
}

File: golang/leetcode/31.go
package leetcode

func nextPermutation(nums []int)  {
	n := len(nums)
	swap := func(i, j int){
		nums[i], nums[j] = nums[j], nums[i]
	}
	reverse := func(l, r int){
		for l < r{
			swap(l, r)
			l += 1
			r -= 1
		}
	}
	i := n-2
	for ; i>=0 && nums[i] >= nums[i+1]; i -=1{}
	if i >= 0{
		for j:=n-1; j>i; j -= 1 {
			if nums[j] > nums[i]{
				swap(i, j)
				break
			}
		}
		reverse(i+1, n-1)
	}else{
		reverse(0, n-1)
	}
}
